components/script/dom/permissionstatus.rs

```diff
@@ -0,0 +1,189 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use core::clone::Clone;
+use dom::bindings::cell::DOMRefCell;
+use dom::bindings::codegen::Bindings::BluetoothPermissionResultBinding::BluetoothPermissionDescriptor;
+use dom::bindings::codegen::Bindings::EventHandlerBinding::EventHandlerNonNull;
+use dom::bindings::codegen::Bindings::PermissionStatusBinding::{self, PermissionDescriptor, PermissionName};
+use dom::bindings::codegen::Bindings::PermissionStatusBinding::{PermissionState, PermissionStatusMethods};
+use dom::bindings::error::ErrorResult;
+use dom::bindings::js::Root;
+use dom::bindings::reflector::{DomObject, reflect_dom_object};
+use dom::bindings::trace::JSTraceable;
+use dom::eventtarget::EventTarget;
+use dom::globalscope::GlobalScope;
+use heapsize::HeapSizeOf;
+use js::jsapi::{JS_GetIsSecureContext, JSObject, JSTracer};
+use js::rust::get_object_compartment;
+
+// Enum for storing different type of PermissionDescriptors in the same type.
+#[derive(JSTraceable, HeapSizeOf)]
+pub enum PermissionDescriptorType {
+    // TODO(zakorgy): Finish this list.
+    Undefined,
+    Default(PermissionDescriptor),
+    Bluetooth(BluetoothPermissionDescriptor),
+}
+
+// https://w3c.github.io/permissions/#permissionstatus
+#[dom_struct]
+pub struct PermissionStatus {
+    eventtarget: EventTarget,
+    state: DOMRefCell<PermissionState>,
+    query: DOMRefCell<PermissionDescriptorType>,
+}
+
+impl PermissionStatus {
+    pub fn new_inherited() -> PermissionStatus {
+        PermissionStatus {
+            eventtarget: EventTarget::new_inherited(),
+            state: DOMRefCell::new(PermissionState::Denied),
+            query: DOMRefCell::new(PermissionDescriptorType::Undefined),
+        }
+    }
+
+    pub fn new(global: &GlobalScope) -> Root<PermissionStatus> {
+        reflect_dom_object(box PermissionStatus::new_inherited(),
+                           global,
+                           PermissionStatusBinding::Wrap)
+    }
+
+    pub fn set_query(&self, permission_descriptor: PermissionDescriptorType) {
+        *self.query.borrow_mut() = permission_descriptor;
+    }
+
+    pub fn get_query(&self) -> &DOMRefCell<PermissionDescriptorType> {
+        &self.query
+    }
+
+    // https://w3c.github.io/permissions/#create-a-permissionstatus
+    pub fn create_from_descriptor(global: &GlobalScope,
+                                  permission_desc: PermissionDescriptorType)
+                                  -> Root<PermissionStatus> {
+        let permission_status = PermissionStatus::new(global);
+        permission_status.set_query(permission_desc);
+        permission_status
+    }
+
+    // https://w3c.github.io/permissions/#boolean-permission-query-algorithm
+    pub fn permission_query(&self, permission_desc: &PermissionDescriptorType) {
+        // Step 1.
+        *self.state.borrow_mut() =
+            get_descriptors_permission_state(permission_desc, self.eventtarget.reflector().get_jsobject().get());
```

Comment:
No it's 114 character long. Also there is a mark in `./mach test-tidy does not report any errors` .